### PR TITLE
[FLINK-29713] Kubernetes operator should restart failed jobs

### PR DESCRIPTION
**Pull request by pvary**
## What is the purpose of the change

It would be good to have the possibility to restart the Flink Application if it goes to `FAILED` state.
This could be used to restart, and reconfigure the job dynamically in the application `main` method if the current application can not handle the incoming data.

## Brief change log

  - Creates a new configuration `cluster.restart.failed` with the default `false`
  - If the Application state is `FAILED` and the configuration is set to `true` then the application is removed and redeployed

## Verifying this change
This change added tests and can be verified as follows:
  - Added tests where the configuration is turned on, and the application is redeployed

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changes to the `CustomResourceDescriptors`: no
  - Core observer or reconciler logic that is regularly executed: yes

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

### CODE REVIEW

File: docs/content/docs/custom-resource/job-management.md [hunk 1]
```diff
@@ -241,6 +241,21 @@ In order this feature to work one must enable [recovery of missing job deploymen
...
+
+The operator can restart a failed Flink cluster deployment. This could be useful in cases when the job main task is
+able to reconfigure the job to handle these failures.
+
+For example a job could dynamically create the DAG based on some job configuration which job configuration could
+change over time. When a task detects a record which could not be handled with the current configuration then the task
+should throw a `SuppressRestartsException` to fail the job. If `kubernetes.operator.cluster.restart.failed` is set to 
+`true` (default: `false`) then the operator detects the failed job and restarts it. When the job restarts then it reads
+the new job configuration and creates the new DAG based on this new configuration. The new deployment could handle the
+incoming records and no manual intervention is needed.
```
Comment (gyfora) [inline docs/content/docs/custom-resource/job-management.md:254]: Let's remove this paragraph, this is very use-case specific.

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java [hunk 2]
```diff
@@ -339,4 +339,11 @@ public static String operatorConfigKey(String key) {
                     .withDescription(
                             "The threshold which is checked against job restart count within a configured window. "
                                     + "If the restart count is reaching the threshold then full cluster restart is initiated.");
+
+    @Documentation.Section(SECTION_DYNAMIC)
+    public static final ConfigOption<Boolean> OPERATOR_CLUSTER_RESTART_FAILED =
+            operatorConfig("cluster.restart.failed")
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java:345]: this should be:

```
job.restart.failed
```

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconciler.java [hunk 3]
```diff
@@ -262,26 +264,44 @@ public boolean reconcileOtherChanges(
             return true;
         }
 
+        if (JobStatus.valueOf(deployment.getStatus().getJobStatus().getState()) == JobStatus.FAILED
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconciler.java:267]: Maybe we could move this logic up one level to the `AbstractJobReconciler` this logic and config makes sense for both Applications and SessionJobs

**Review by gyfora — CHANGES_REQUESTED** [review]

File: docs/content/docs/custom-resource/job-management.md (hunk 1 above)
Comment (pvary) [reply docs/content/docs/custom-resource/job-management.md]: Done

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java (hunk 2 above)
Comment (pvary) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/config/KubernetesOperatorConfigOptions.java]: Done

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconciler.java (hunk 3 above)
Comment (pvary) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/ApplicationReconciler.java]: Done

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractJobReconciler.java [hunk 4]
```diff
@@ -238,8 +240,34 @@ protected void rollback(CR resource, Context<?> ctx, Configuration observeConfig
...
+            throws Exception {
+        LOG.info("Resubmitting Flink Cluster deployment...");
+        SPEC specToRecover = ReconciliationUtils.getDeployedSpec(deployment);
+        restoreJob(
+                deployment,
+                specToRecover,
+                deployment.getStatus(),
+                ctx,
+                observeConfig,
+                requireHaMetadata);
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractJobReconciler.java:270]: The `resubmitJmDeployment` name is not correct here. Maybe it's better to simply inline the `restoreJob` call in the if branch if it's only called in one place

Comment (gyfora) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractJobReconciler.java:270]: Or just rename it to `resubmintJob` and modify the logging accordingly

Comment (pvary) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractJobReconciler.java]: I like this. Renamed to `resubmitJob`, and modified the log message accordingly

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractJobReconciler.java [hunk 5]
```diff
@@ -253,4 +281,14 @@ public boolean reconcileOtherChanges(
...
+
+    /**
+     * Removes a failed job.
+     *
+     * @param resource The failed job.
+     * @param observeConfig Observe configuration.
+     * @throws Exception Error during cancellation.
+     */
+    protected abstract void removeFailedJob(
+            CR resource, Context<?> ctx, Configuration observeConfig) throws Exception;
```
Comment (gyfora) [inline flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractJobReconciler.java:293]: I suggest we rename this to `cleanupAfterFailedJob`

**Review by gyfora — APPROVED** [review]
One minor comment about method naming, otherwise looks good :)

File: flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractJobReconciler.java (hunk 5 above)
Comment (pvary) [reply flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractJobReconciler.java]: Done

**Review by gyfora — APPROVED** [review]
Very nice job @pvary!
I will merge this later today after we fixed the flaky test for the CI

**Comment by jeesmon** [comment]
@pvary Can this be configured per FlinkDeployment or is it global config for all Deployments managed by the operator? Thanks!

**Comment by gyfora** [comment]
> @pvary Can this be configured per FlinkDeployment or is it global config for all Deployments managed by the operator? Thanks!

@jeesmon this can be configured on a per CR basis (it is in dynamic section, thats the hint)

**Comment by jeesmon** [comment]
@gyfora Just to confirm, so just need to add it to the `spec.flinkConfiguration` if I want to enable it just for a specific `FlinkDeployment`?

**Comment by gyfora** [comment]
@jeesmon yes, same as with any other config

**Pull request merged**